python3Packages.relaxBuildSystemRequiresHook: init (#331315)
NixOS/nixpkgs: doc/languages-frameworks/python.section.md

- `pythonRelaxDepsHook` will relax Python dependencies restrictions for the package.
  See [example usage](#using-pythonrelaxdepshook).
- `pythonRemoveBinBytecode` to remove bytecode from the `/bin` folder.

> - `relaxBuildSystemRequiresHook` to relax build-system restrictions defined in PEP518 for the package.

Reviewer: Hmm most people are not familiar with the different PEPs, so how about:

```suggestion
- `relaxBuildSystemRequiresHook` to relax build-system restrictions defined in projects' `pyproject.toml`.
```

The `pythonRelaxDepsHook` has no effect on build time dependencies, such as
those specified in `build-system`. If a package requires incompatible build
time dependencies, consider using `relaxBuildSystem` or `removeBuildSystem`.

#### Using relaxBuildSystemRequiresHook {#using-relaxbuildsystemrequireshook}

`relaxBuildSystemRequiresHook` has the same function as `pythonRelaxDepsHook`,
but it affects `build-system` instead of `dependencies`.

For example, given the following `pyproject.toml` file:

```toml
[build-system]
requires = [
  "pkg1<1.0",
  "pkg2",
  "pkg3>=1.0,<=2.0",
]
```

we can do:

```nix
{
  relaxBuildSystem = [
    "pkg1"
    "pkg3"
  ];
  removeBuildSystem = [
    "pkg2"
  ];
}
```

which would result in the following `pyproject.toml` file:

```toml
[build-system]
requires = [
  "pkg1",
  "pkg3",
]
```

In general you should always use `relaxBuildSystem`, because `removeBuildSystem`
will cause silent build errors. However `removeBuildSystem` may
still be useful in exceptional cases, and also to remove dependencies wrongly
declared by upstream (for example, declaring `pytest` as a build dependency

> instead of a test dependency).

Reviewer 2: I think also mentioning `numpy_2` here is a good idea, since they [explicitly reccomend](https://numpy.org/devdocs/dev/depending_on_numpy.html#numpy-2-0-specific-advice) that you add `numpy>=2` to `requires` even when compatible with v1

The `pythonRelaxDepsHook` has no effect on build time dependencies, such as
those specified in `build-system`. If a package requires incompatible build

> time dependencies, consider using `relaxBuildSystem` or `removeBuildSystem`.

Reviewer: The `Deps` suffix missing is confusing IMO, how about naming these attributes:

```suggestion
time dependencies, consider using `pythonRelaxBuildDeps` or `pythonRemoveBuildDeps`.
```

#### Using relaxBuildSystemRequiresHook {#using-relaxbuildsystemrequireshook}

> `relaxBuildSystemRequiresHook` has the same function as `pythonRelaxDepsHook`,

Reviewer: Similarly, I'd suggest a different name, analogue to the existing attributes:

```suggestion
`pythonRelaxBuildDepsHook` has the same function as `pythonRelaxDepsHook`,
```

* Only unversioned attributes (e.g. `pydantic`, but not `pypdantic_1`) can be included in `dependencies`,
  since due to `PYTHONPATH` limitations we can only ever support a single version for libraries
  without running into duplicate module name conflicts.
* The version restrictions of `dependencies` can be relaxed by [`pythonRelaxDepsHook`](#using-pythonrelaxdepshook)

>   and of `build-system` can be relaxed by [`relaxBuildSystemRequiresHook`](#using-relaxbuildsystemrequireshook).

Reviewer: ```suggestion
  and of `build-system` can be relaxed by [`pythonRelaxBuildDepsHook`](#using-pythonRelaxBuildDepsHook).
```
